### ISSUE TITLE
Android support

### ISSUE THREAD

**Issue by dobkeratops** [PistonDevelopers/piston#116]
Interesting; 
have you considered tablet/phone-friendliness in the framework? ( e.g. android, iOS )
(how do i specify 'discussion' label?)

So far I've been doing cross platform (iOS/android/windows) in C++, with some simple functions that can be called from a platform specific stub (init, update, render) ..but also 'display_init' and 'display_close') for android.
 
using 'extern "C"' I could just slot Rust code into the same desktop/android framework as my existing C/C++ work.

I haven't attempted to make any "android NDK application stub" in Rust as yet... I also haven't looked into GLFW for android. Perhaps the ambition of this project would be to do a complete back-end on every platform , natively in rust, eventually ..

supposedly rust tries to emit the same name-mangling as C++. however C++ doesn't have  a standard for that. extern "C" loses the elegance of either language (rust or c++) but does allow crossover which is nice.

I think it would be pretty interesting to have a simple cross platform 2d library (including some sort of input) designed for mobiles from the ground up.. menus that easily map onto touchscreen/desktop and so on. This is not trivial to design , and you'd have to consider the touchscreen as the "lowest common denominator" really, but add keyboard and mouse buttons as intelligent shortcuts (so many laptops have touch screens now); of course you wouldn't want to compromise the ability to make something that works well on a specific system.

**Comment by beatgammit** [comment]
@bvssvni - Is this viable now? I'm really interested in playing with game dev on mobile with rust.

**Comment by bvssvni** [comment]
@beatgammit Yeah, it is time to think about it again. There has been a few experiments but I have not tried it myself.

**Comment by mhsjlw** [comment]
Any progress on this, I'd love to use this to make a cross-platform Linux/Mac/Android game!

**Comment by opensourcegeek** [comment]
Any updates on this please? I'm toying with a game idea and it would be great to be able to run on android/ios. Thanks for wonderful work so far.

**Comment by dylanhart** [comment]
I was able to get an older version of Piston able to be cross compiled to android using [`cargo-apk`](https://github.com/tomaka/android-rs-glue).

I create a window with `glutin` that requests OpenGLES as a fallback. I then wrap it in a modified version of the wrapper from `glutin_window` (that takes an existing window). Then, I swap out graphics backends between `opengl_graphics` and [`opengles_graphics`](https://crates.io/crates/opengles_graphics) using `#[cfg(target_os = "android")]`.

This is using `graphics` version v0.21.0 to work with `opengles_graphics` which is targeting this older version.

**Comment by OptimisticPeach** [comment]
Unfortunately `opengles_graphics` is 3 years old and ancient and deprecated. Is there another kind backend for opengles that can be used?
